### ISSUE TITLE
False negative: `TypeError: metaclass conflict` when using Protocols and metaclasses

### ISSUE THREAD

**Issue by Daraan** [microsoft/pyright#8982]
**Describe the bug**
The following code raises a `TypeError: metaclass conflict` that is not reported

Python version 3.10 & 3.11, others not checked

```python
# pyright: strict
from typing import TYPE_CHECKING, Protocol

class TheProto(Protocol): ...
class TheMeta(type):  ...

# Raises error but does not report it statically
class HasRuntimeError(TheProto, metaclass=TheMeta): ...  # <-- error
```

**VS Code extension or command-line**
CLI: pyright 1.1.380
and 1.1.378 from Pylance

---

Some hacks to work around it at runtime

```python

# Some hacks around to be valid runtime
class RuntimeOK(TheProto if TYPE_CHECKING else object, metaclass=TheMeta): ...

class HackProto(Protocol if TYPE_CHECKING else object): ...
class RuntimeOK2(HackProto if TYPE_CHECKING else object, metaclass=TheMeta): ...
```

**Comment by erictraut** [comment]
There are a couple of issues here:
1. The `Protocol` symbol is defined as a special form in typeshed stubs. It isn't defined as a class with a metaclass. A type checker could theoretically hard-code this knowledge for `Protocol`. The `_ProtocolMeta` class definition is included in the `typing.pyi` stub.
2. Many stdlib classes defined in the typeshed stubs claim to derive from `Protocol` when they actually do not. For example, `str` is defined to have `Sequence` as a base class which is defined to have `Collection` as a base class which is defined to derive from `Protocol`. At runtime, `Protocol` is not in the MRO for `str` and `_ProtocolMeta` is not its metaclass. The typeshed stubs are kind of lying in this case. If I correctly model the metaclass for the `Protocol` class, it results in many false positive errors. For example, `class StrEnum(str, Enum)` would appear to have a metaclass conflict when in reality it does not.

Mypy has apparently found a way to model this such that it generates a "metaclass conflict" error for the code above but doesn't generate false positives. I'm not sure what technique it's using to thread this needle. I may need to do some spelunking in the mypy code base.

@JelleZijlstra, do you have any insights here?

**Comment by JelleZijlstra** [comment]
I don't have too much to add and I'm not sure what special-casing mypy does to cover this case. Here is a related typeshed issue: python/typeshed#1595.

It might be nice if we could write the typeshed stub for str like:
```
@Sequence[str].register
class str: ...
```
But that would require spec changes and coordination with other type checkers.

**Comment by erictraut** [comment]
I'm going to close this issue.

Given that the typeshed stubs lie about many classes deriving from `Protocol`, implementing this check would result in many false positives. I've looked at mypy's implementation of this check, and it appears to be incorrect — or at least inconsistent, so I'm not inclined to replicate its behavior.

While this case is arguably a false negative, I don't think this is a frequent source of bugs, and when this bug does occur it is not at all subtle. The interpreter will inform you of the problem the first time you attempt to run the code.

I like Jelle's suggestion for registering classes as virtual protocols, but as he points out, this will require a spec update and coordination with other type checkers. @Daraan, if you'd like to spearhead such an effort, you're welcome to do so.
